ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java

```diff
@@ -261,11 +284,13 @@
   @Override
   public boolean hasSucceeded() {
     boolean hasSucceeded;
-    if (routerOperation == RouterOperation.PutOperation && routerConfig.routerPutUseDynamicSuccessTarget) {
-      hasSucceeded =
-          succeededCount >= Math.max(totalReplicaCount - disabledCount - 1, routerConfig.routerPutSuccessTarget);
+    if (routerOperation == RouterOperation.PutOperation && routerConfig.routerPutUseDynamicSuccessTarget
+        && inFlightReplicaType == ReplicaType.DISK_BACKED) {
+      // this logic only applies to disk replicas where the quorum can change during replica movement
+      int dynamicSuccessTarget = Math.max(totalReplicaCount - disabledCount - 1, routerConfig.routerPutSuccessTarget);
+      hasSucceeded = diskSucceededCount >= dynamicSuccessTarget;
     } else {
-      hasSucceeded = succeededCount >= successTarget;
+      hasSucceeded = diskSucceededCount >= diskSuccessTarget || cloudSucceededCount >= cloudSuccessTarget;
```

Comment:
This logic looks correct.  I'm not convinced about the hasFailed logic, seems like that needs to use a similar OR condition.